build_modules/lib/src/platform.dart

```diff
@@ -2,110 +2,53 @@
+///   different).
+///
+/// Typically these should correspond to `libraries.json` files in the SDK.
+///
+/// New platforms should be created with [register], and can later be
+/// fetched by name using the [DartPlatform.byName] static method.
 class DartPlatform {
+  /// A list of all registered platforms by name, populated by
+  /// [register].
+  static final _platformsByName = <String, DartPlatform>{};
+
   final List<String> _supportedLibraries;
 
   final String name;
 
-  static const dartdevc = DartPlatform._('dartdevc', [
-    'async',
-    'collection',
-    'convert',
-    'core',
-    'developer',
-    'html',
-    'html_common',
-    'indexed_db',
-    'js',
-    'js_util',
-    'math',
-    'svg',
-    'typed_data',
-    'web_audio',
-    'web_gl',
-    'web_sql',
-  ]);
-  static const dart2js = DartPlatform._('dart2js', [
-    'async',
-    'collection',
-    'convert',
-    'core',
-    'developer',
-    'html',
-    'html_common',
-    'indexed_db',
-    'js',
-    'js_util',
-    'math',
-    'svg',
-    'typed_data',
-    'web_audio',
-    'web_gl',
-    'web_sql',
-  ]);
-  static const dart2jsServer = DartPlatform._('dart2js_server', [
-    'async',
-    'collection',
-    'convert',
-    'core',
-    'developer',
-    'js',
-    'js_util',
-    'math',
-    'typed_data',
-  ]);
-  static const flutter = DartPlatform._('flutter', [
-    'async',
-    'cli',
-    'collection',
-    'convert',
-    'core',
-    'developer',
-    'io',
-    'isolate',
-    'math',
-    'nativewrappers',
-    'profiler',
-    'typed_data',
-    'ui',
-    'vmservice_io',
-  ]);
-  static const vm = DartPlatform._('vm', [
-    'async',
-    'cli',
-    'collection',
-    'convert',
-    'core',
-    'developer',
-    'io',
-    'isolate',
-    'math',
-    'mirrors',
-    'nativewrappers',
-    'profiler',
-    'typed_data',
-    'vmservice_io',
-  ]);
-
-  factory DartPlatform(String name) {
-    var platform = _byName[name];
-    if (platform == null) {
-      throw ArgumentError('Unrecognized paltform $name, the recognized '
-          'platforms are ${_byName.keys.join(', ')}.');
+  /// Returns a [DartPlatform] instance by name.
+  ///
+  /// Throws an [UnrecognizedDartPlatformException] if [name] has not been
+  /// registered with [DartPlatform.register].
+  static DartPlatform byName(String name) {
+    var platform = _platformsByName[name];
+    if (platform != null) return platform;
+    throw UnrecognizedDartPlatformException(name);
+  }
+
+  /// Registers a new [DartPlatform].
+  ///
+  /// Throws a [DartPlatformAlreadyRegisteredException] if [name] has already
+  /// been registered by somebody else.
+  static DartPlatform register(String name, List<String> supportedLibraries) {
+    if (_platformsByName.containsKey(name)) {
+      throw DartPlatformAlreadyRegisteredException(name);
     }
+
+    var platform = DartPlatform._(name, List.of(supportedLibraries));
+    _platformsByName[name] = platform;
```

Comment:
[optional] Assignment also evaluates to the value.

```dart
return _platformsByName[name] = DartPlatform._(name, List.of(supportedLibraries));
```

Reply:
just looks weird to me but ok :)